plugins/artifacts/artifacts.go

```diff
@@ -1,143 +1,315 @@
+		// Abort, if task context is cancelled
+		if tp.context.Err() != nil {
+			return
+		}
+		a := tp.artifacts[i]
+		if a.Expires.IsZero() {
+			a.Expires = tp.context.TaskInfo.Expires
+		}
+		switch a.Type {
+		case typeFile:
+			tp.processFile(result, a)
+		case typeDirectory:
+			tp.processDirectory(result, a)
+		}
+	})
+	debug("Artifacts extracted and uploaded")
+
+	// Find error condition
+	var err error
+	if len(tp.errors) > 0 {
+		err = runtime.MergeMalformedPayload(tp.errors...)
+	}
+	if tp.nonFatalErr.Get() {
+		err = runtime.ErrNonFatalInternalError
+	}
+	if tp.fatalErr.Get() {
+		err = runtime.ErrFatalInternalError
+	}
+	return !tp.failed.Get(), err
 }
 
-func (tp *taskPlugin) Stopped(result engines.ResultSet) (bool, error) {
-	nonFatalErrs := []runtime.MalformedPayloadError{}
+const (
+	reasonFileMissing     = "file-missing-on-worker"
+	reasonInvalidResource = "invalid-resource-on-worker"
+	reasonTooLarge        = "too-large-file-on-worker"
+)
 
-	for _, artifact := range tp.artifacts {
-		// If expires is set to this time it's the default value
-		if artifact.Expires.IsZero() {
-			artifact.Expires = tp.context.TaskInfo.Expires
-		}
-		switch artifact.Type {
-		case "directory":
-			err := result.ExtractFolder(artifact.Path, tp.createUploadHandler(artifact.Name, artifact.Expires))
-			if err != nil {
-				if tp.errorHandled(artifact.Name, artifact.Expires, err) {
-					nonFatalErrs = append(nonFatalErrs, runtime.NewMalformedPayloadError(err.Error()))
-					continue
-				}
-				return false, err
-			}
-		case "file":
-			fileReader, err := result.ExtractFile(artifact.Path)
-			if err != nil {
-				if tp.errorHandled(artifact.Name, artifact.Expires, err) {
-					nonFatalErrs = append(nonFatalErrs, runtime.NewMalformedPayloadError(err.Error()))
-					continue
-				}
-				return false, err
-			}
-			err = tp.attemptUpload(fileReader, artifact.Path, artifact.Name, artifact.Expires)
-			if err != nil {
-				return false, err
-			}
+func (tp *taskPlugin) processFile(result engines.ResultSet, a artifact) {
+	debug("extracting file from path: %s", a.Path)
+	r, err := result.ExtractFile(a.Path)
+	// Always close reader, if one is returned
+	defer func() {
+		if r != nil {
+			r.Close()
 		}
+	}()
+
+	// If feature isn't supported this is malformed-payload
+	if err == engines.ErrFeatureNotSupported {
+		e := runtime.NewMalformedPayloadError(
+			"Artifact extraction is not supported in current configuration of this workerType",
+		)
+		tp.mErrors.Lock()
+		tp.errors = append(tp.errors, e)
+		tp.mErrors.Unlock()
+		return
 	}
 
-	if len(nonFatalErrs) > 0 {
-		// Only report an exception if the command executed successfully.
-		// The logic behind this it is that no artifact upload failure is
-		// expected for a succeeded run, but failed tasks might be missing
-		// some artifacts.
+	// If resource isn't found, task should fail and we print a message to task log
+	if err == engines.ErrResourceNotFound {
```

Comment:
Wouldn't this be `err == engines.ErrResourceNotFound && result.Success()`?
If the task has failed, it is expected some artifacts are missing, no reason to fail the plugin.